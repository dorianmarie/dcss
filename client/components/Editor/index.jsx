import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { NavLink, withRouter } from 'react-router-dom';
import {
  Button,
  Dropdown,
  Header,
  Icon,
  Menu,
  Modal,
  Segment,
  Title
} from '@components/UI';
import Storage from '@utils/Storage';
import EditorMenu from '@components/EditorMenu';
import { notify } from '@components/Notification';
import ScenarioEditor from '@components/ScenarioEditor';
import ScenarioStatusMenuItem from '@components/EditorMenu/ScenarioStatusMenuItem';
import Scenario from '@components/Scenario';
import Username from '@components/User/Username';
// Review is presently not in use, but will be introduced along
// with more features for scenario collaboration.
// import Review from './Review';
import Slides from '@components/Editor/Slides';
import Identity from '@utils/Identity';
import { makeDefaultDescription } from './scenario';
import {
  copyScenario,
  deleteScenario,
  getScenario,
  setScenario,
  endScenarioLock
} from '@actions/scenario';
import { getPersonas } from '@actions/persona';
import { getUsers } from '@actions/users';
import './editor.css';

class Editor extends Component {
  constructor(props) {
    super(props);

    this.copyScenario = this.copyScenario.bind(this);
    this.deleteScenario = this.deleteScenario.bind(this);
    this.getAllTabs = this.getAllTabs.bind(this);
    this.getPostSubmitCallback = this.getPostSubmitCallback.bind(this);
    this.getSubmitCallback = this.getSubmitCallback.bind(this);
    this.getTab = this.getTab.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onClickScenarioAction = this.onClickScenarioAction.bind(this);
    // this.onBeforeUnload = this.onBeforeUnload.bind(this);
    this.setActiveView = this.setActiveView.bind(this);
    this.updateScenario = this.updateScenario.bind(this);

    let noPersistedView = true;
    let {
      activeTab,
      activeSlideIndex,
      isCopyScenario,
      isNewScenario,
      match,
      scenarioId
    } = this.props;

    if (!scenarioId) {
      scenarioId = isNewScenario ? 'new' : Number(match.params.id);
    }

    if (isNewScenario) {
      activeTab = 'scenario';
    }

    if (!isCopyScenario && !isNewScenario) {
      this.sessionKey = `editor/${scenarioId}`;

      let persisted = Storage.get(this.sessionKey, {
        activeTab: 'scenario',
        activeSlideIndex
      });

      // These have already been declared as let bindings above
      // but we may override those values here, with whatever
      // was persisted for this scenario
      ({ activeTab, activeSlideIndex } = persisted);

      noPersistedView = false;
    }

    this.state = {
      isReady: false,
      activeSlideIndex,
      activeTab,
      scenarioId,
      tabs: null
    };

    if (isCopyScenario) {
      this.copyScenario(scenarioId);
    }

    if (!isNewScenario) {
      if (noPersistedView) {
        this.state.activeTab = 'slides';
      }
    }
  }

  async componentDidMount() {
    if (!this.props.isNewScenario) {
      const scenario = await this.props.getScenario(this.props.scenarioId, {
        lock: true
      });

      const { scenarioUser } = this.props;
      if (scenario && !scenario.lock) {
        this.props.history.push('/scenarios');
        return;
      }

      // The viewing user might be a super admin that is not
      // a collaborator on this scenario.
      if (scenarioUser) {
        if (
          scenario.lock.user_id === scenarioUser.id &&
          scenarioUser.is_reviewer
        ) {
          await this.props.endScenarioLock(scenario.id);
        }
      }
    }

    await this.props.getUsers();

    if (!this.props.personas.length) {
      await this.props.getPersonas();
    }

    this.setState(state => {
      const { scenarioUser } = this.props;

      return {
        isReady: true,
        activeTab:
          scenarioUser && scenarioUser.is_reviewer ? 'review' : state.activeTab,
        tabs: this.getAllTabs(state.scenarioId)
      };
    });
  }

  onClick(e, { name: activeTab }) {
    this.setState({ activeTab });
    const { scenarioId } = this.state;
    const activeNonZeroSlideIndex = Number(
      this.props.match.params.activeNonZeroSlideIndex ||
        this.props.match.params.activeRunSlideIndex
    );

    const pathname = `/editor/${scenarioId}/${activeTab}/${activeNonZeroSlideIndex}`;

    this.props.history.push(pathname);

    Storage.set(this.sessionKey, {
      activeSlideIndex: activeNonZeroSlideIndex,
      activeTab
    });
  }

  setActiveView({ activeTab, activeSlideIndex }) {
    const { scenarioId } = this.state;
    /*

        High likelihood that this is buggy.


     */

    let activeNonZeroSlideIndex = activeSlideIndex + 1 || 1;
    let pathname;

    if (activeTab === 'preview') {
      activeNonZeroSlideIndex = activeSlideIndex;
      pathname = `/editor/${scenarioId}/${activeTab}/${activeNonZeroSlideIndex}`;
      this.setState({ activeTabSlideIndex: activeNonZeroSlideIndex });
    } else {
      pathname = `/editor/${scenarioId}/${activeTab}/${activeNonZeroSlideIndex}`;
      this.setState({ activeSlideIndex });
    }

    this.props.history.push(pathname);
  }

  onClickScenarioAction(event, data) {
    if (data.name === 'save-scenario') {
      this.updateScenario();
    }

    if (data.name === 'save-status') {
      this.updateScenario({ status: data.id });
    }
  }

  async copyScenario(id) {
    if (!id) {
      return;
    }

    const scenario = await this.props.copyScenario(id);

    if (!scenario) {
      notify({ type: 'error', message: 'Error saving copy.' });
      return;
    }

    // Hard redirect to clear all previous state from the editor.
    // DO NOT USE this.props.history.push(...)
    location.href = `/editor/${scenario.id}`;
  }

  async deleteScenario(scenario) {
    await this.props.deleteScenario(scenario.id);
    this.props.history.push('/scenarios/');
  }

  // TODO: Move to own async action
  async updateScenario(updates = {}) {
    // NOTE: this is to support saving the whole
    //       scenario when clicking the [save icon]
    //       that's displayed via EditorMenu.
    const {
      author,
      categories,
      consent,
      description,
      finish,
      labels,
      personas,
      status,
      title
    } = this.props.scenario;

    const data = {
      author,
      categories,
      consent,
      description,
      finish,
      labels,
      personas,
      status,
      title
    };

    Object.assign(data, updates);

    const submitCallback = this.getSubmitCallback();
    // TODO: Move to own async action
    const response = await (await submitCallback(data)).json();

    if (response.error) {
      notify({ type: 'error', message: response.message });
    } else {
      this.props.setScenario({
        ...this.props.scenario,
        ...response.scenario
      });
      notify({ type: 'success', message: 'Scenario saved' });
    }
  }

  getTab(name, scenarioId) {
    const { setActiveView } = this;
    switch (name) {
      case 'scenario':
        // TODO: Eliminate scenarioId
        return (
          <ScenarioEditor
            scenarioId={scenarioId}
            submitCB={this.getSubmitCallback()}
            postSubmitCB={this.getPostSubmitCallback()}
          />
        );
      case 'slides':
        // TODO: Eliminate scenarioId
        return (
          <Slides
            setActiveSlide={activeSlideIndex =>
              setActiveView({
                activeSlideIndex,
                activeTab: 'slides'
              })
            }
            scenarioId={scenarioId}
            scenario={this.props.scenario}
          />
        );
      case 'preview':
        // TODO: Eliminate scenarioId
        return (
          <Scenario
            key={scenarioId}
            setActiveSlide={activeSlideIndex =>
              setActiveView({
                activeSlideIndex,
                activeTab: 'preview'
              })
            }
            scenarioId={scenarioId}
            scenario={this.props.scenario}
          />
        );
      case 'review':
        return (
          <Scenario
            key={scenarioId}
            setActiveSlide={activeSlideIndex =>
              setActiveView({
                activeSlideIndex,
                activeTab: 'preview'
              })
            }
            scenarioId={scenarioId}
            scenario={this.props.scenario}
          />
        );
      // return (
      //   <Review scenario={this.props.scenario} />
      // );
      default:
        return null;
    }
  }

  getAllTabs() {
    const { scenario, scenarioUser } = this.props;

    switch (scenario.id) {
      case undefined:
        return {
          scenario: this.getTab('scenario', 'new')
        };
      default: {
        const authorTabs = {
          scenario: this.getTab('scenario', scenario.id),
          slides: this.getTab('slides', scenario.id),
          preview: this.getTab('preview', scenario.id)
        };

        const reviewerTabs = {
          review: this.getTab('review', scenario.id)
        };

        return scenarioUser && scenarioUser.is_reviewer
          ? reviewerTabs
          : authorTabs;
      }
    }
  }

  getSubmitCallback() {
    let endpoint, method;
    const { isNewScenario, scenarioId } = this.props;

    if (isNewScenario) {
      endpoint = '/api/scenarios';
      method = 'POST';
    } else {
      endpoint = `/api/scenarios/${scenarioId}`;
      method = 'PUT';
    }

    // TODO: Move to own async action
    return scenario => {
      if (method === 'POST' && !scenario.description) {
        scenario.description = makeDefaultDescription(scenario);
      }

      return fetch(endpoint, {
        method,
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(scenario)
      });
    };
  }

  getPostSubmitCallback() {
    const { history, isCopyScenario, isNewScenario } = this.props;
    const { setActiveView } = this;

    if (isCopyScenario || isNewScenario) {
      return scenario => {
        history.push(`/editor/${scenario.id}`);
        this.setState({ scenarioId: scenario.id }, () => {
          setActiveView({ activeTab: 'slides', activeSlideIndex: 1 });
        });
      };
    }

    return null;
  }

  render() {
    const { isReady, scenarioId } = this.state;

    if (this.props.isCopyScenario) {
      return null;
    }

    if (!this.state.tabs) {
      return null;
    }

    if (!isReady) {
      return null;
    }

    const { scenario, scenarioUser, user, usersById } = this.props;
    const isNotNew = scenarioId !== 'new';

    const menuItemScenarioStatus = scenario.status !== undefined && (
      <ScenarioStatusMenuItem
        tabIndex="0"
        key="menu-item-scenario-status"
        name="Set scenario status"
        status={scenario.status}
        onChange={this.onClickScenarioAction}
      />
    );

    const menuItemScenarioCopy = isNotNew ? (
      <Menu.Item.Tabbable
        key="menu-item-scenario-run"
        name="Copy this scenario"
        onClick={() => this.copyScenario(scenario.id)}
      >
        <Icon name="copy outline" />
      </Menu.Item.Tabbable>
    ) : null;

    const menuItemScenarioRun = isNotNew ? (
      <Menu.Item.Tabbable
        key="menu-item-scenario-run"
        name="Run this scenario"
        onClick={() => {
          this.props.history.push(`/run/${scenarioId}/slide/0`);
        }}
      >
        <Icon name="play" />
      </Menu.Item.Tabbable>
    ) : null;

    const menuItemScenarioUnlock = isNotNew ? (
      <Menu.Item.Tabbable
        key="menu-item-scenario-unlock"
        name="Unlock this scenario"
        onClick={() => {
          this.props.endScenarioLock(this.props.scenario.id);
        }}
      >
        <Icon name="unlock" />
      </Menu.Item.Tabbable>
    ) : null;

    const menuItemsForAttachedTabularBar = Object.keys(this.state.tabs).map(
      tabType => {
        return (
          <Menu.Item.Tabbable
            key={tabType}
            name={tabType}
            active={this.state.activeTab === tabType}
            onClick={this.onClick}
          />
        );
      }
    );

    let modal = null;

    if (isNotNew) {
      const modalProps = {
        open: false
      };
      let header = '';
      let content = '';
      const hasLock = scenario.lock && scenario.lock.user_id !== user.id;

      if (scenarioUser) {
        if (!scenarioUser.is_reviewer && hasLock) {
          modalProps.open = true;

          const lockHolder = usersById[scenario.lock.user_id];
          content = (
            <Fragment>
              This scenario is currently being edited by{' '}
              <Username user={lockHolder} />
            </Fragment>
          );
          setTimeout(() => {
            this.props.history.push('/scenarios/');
          }, 10000);
        }
      } else {
        if (!scenarioUser && !user.is_super) {
          modalProps.open = true;
          content = 'You do not have permission to access this scenario.';
        }
      }

      if (modalProps.open) {
        const ariaLabelledby = Identity.id();
        const ariaDescribedby = Identity.id();
        modal = (
          <Modal.Accessible open={modalProps.open}>
            <Modal
              {...modalProps}
              size="small"
              role="dialog"
              aria-modal="true"
              aria-labelledby={ariaLabelledby}
              aria-describedby={ariaDescribedby}
            >
              <Header id={ariaLabelledby} content={header} icon="lock" />
              <Modal.Content id={ariaDescribedby}>{content}</Modal.Content>
              <Modal.Actions>
                <Button.Group fluid>
                  <Button
                    to="/scenarios/"
                    color="green"
                    content="Go back to scenarios"
                    as={NavLink}
                  />
                  {/*
                  <Button.Or />
                  <Button
                    as={NavLink}
                    to="/scenarios/"
                    color="orange"
                    content="Try again"
                  />
                  */}
                </Button.Group>
              </Modal.Actions>
            </Modal>
          </Modal.Accessible>
        );
      }

      menuItemsForAttachedTabularBar.push(
        <Menu.Menu key="menu-menu-item-tabs-right" position="right">
          <Menu.Item.Tabbable className="editor__righttitle">
            {this.props.scenario.title}
          </Menu.Item.Tabbable>
        </Menu.Menu>
      );
    }

    const right = [
      menuItemScenarioCopy,
      menuItemScenarioRun,
      location.href.includes('localhost') ? menuItemScenarioUnlock : null,
      menuItemScenarioStatus
    ];

    const isNotReviewer = scenarioUser && !scenarioUser.is_reviewer;
    const canDisplayEditorMenu = isNotNew && (isNotReviewer || user.is_super);

    const pageTitle = `Editing "${scenario.title}"`;

    return modal ? (
      modal
    ) : (
      <Fragment>
        <Title content={pageTitle} />

        <Menu attached="top" tabular className="editor__tabmenu">
          {menuItemsForAttachedTabularBar}
        </Menu>

        <Segment attached="bottom" className="editor__content">
          {canDisplayEditorMenu ? (
            <EditorMenu
              className="em__fullwidth"
              type="scenario"
              items={{
                save: {
                  onClick: (...args) => {
                    this.onClickScenarioAction(...args);
                  }
                },
                delete: {
                  onConfirm: () => {
                    this.deleteScenario(scenario);
                  }
                },
                right
              }}
            />
          ) : null}

          {this.state.tabs[this.state.activeTab]}
        </Segment>
      </Fragment>
    );
  }
}

// Note: this silences the warning about "text={}" receiving
// an object, instead of a string.
Dropdown.propTypes.text = PropTypes.any;

Editor.propTypes = {
  activeTab: PropTypes.string,
  activeSlideIndex: PropTypes.number,
  scenarioId: PropTypes.node,
  history: PropTypes.shape({
    goBack: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired
  }).isRequired,
  match: PropTypes.shape({
    path: PropTypes.string,
    params: PropTypes.shape({
      id: PropTypes.node,
      activeNonZeroSlideIndex: PropTypes.node,
      activeRunSlideIndex: PropTypes.node,
      activeSlideIndex: PropTypes.node
    }).isRequired
  }).isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      scenarioCopyId: PropTypes.node
    })
  }).isRequired,
  isCopyScenario: PropTypes.bool,
  isNewScenario: PropTypes.bool,
  endScenarioLock: PropTypes.func.isRequired,
  copyScenario: PropTypes.func.isRequired,
  deleteScenario: PropTypes.func.isRequired,
  getPersonas: PropTypes.func.isRequired,
  personas: PropTypes.array,
  getScenario: PropTypes.func.isRequired,
  setScenario: PropTypes.func.isRequired,
  scenario: PropTypes.object,
  scenarioUser: PropTypes.object,
  user: PropTypes.object,
  usersById: PropTypes.object,
  getUsers: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) => {
  const scenarioId = Number(ownProps.scenarioId);
  const { personas, scenario, user, usersById } = state;
  const scenarioUser = scenario.users.find(u => u.id === user.id);
  return {
    personas,
    scenario,
    scenarioId,
    scenarioUser,
    user,
    usersById
  };
};

const mapDispatchToProps = dispatch => ({
  endScenarioLock: id => dispatch(endScenarioLock(id)),
  copyScenario: id => dispatch(copyScenario(id)),
  deleteScenario: id => dispatch(deleteScenario(id)),
  getPersonas: () => dispatch(getPersonas()),
  getScenario: (id, options = {}) => dispatch(getScenario(id, options)),
  setScenario: params => dispatch(setScenario(params)),
  getUsers: () => dispatch(getUsers())
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Editor)
);
