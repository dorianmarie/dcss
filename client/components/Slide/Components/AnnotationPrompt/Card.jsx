import React from 'react';
import { Icon } from '@components/UI';
import { name } from './meta';

const CardIcon = <Icon className="pen fancy" aria-label={name} />;
const Card = () => (
  <React.Fragment>
    {CardIcon}
    <span className="ser__component-name">{name}</span>
  </React.Fragment>
);

const memo = React.memo(Card);
memo.Icon = CardIcon;
export default memo;
