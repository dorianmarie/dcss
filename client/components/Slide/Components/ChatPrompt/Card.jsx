import React from 'react';
import { Icon } from '@components/UI';
import { name } from './meta';

const Card = () => (
  <React.Fragment>
    <Icon name="chat" aria-label={name} />
    {name}
  </React.Fragment>
);

export default React.memo(Card);
