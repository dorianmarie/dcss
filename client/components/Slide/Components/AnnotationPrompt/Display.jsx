import { type } from './meta';
import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import AudioPlayer from '../AudioPrompt/AudioPlayer';
import Transcript from '../AudioPrompt/Transcript';
import Loading from '@components/Loading';
import {
  Button,
  Card,
  Header,
  Message,
  Pagination,
  Segment
} from '@components/UI';
import PromptRequiredLabel from '../PromptRequiredLabel';
import { getResponse } from '@actions/response';
import { ANSWER_ANNOTATION } from '@hoc/withRunEventCapturing';
import withSocket, {
  RUN_RESPONSE_CREATED,
  RUN_RESPONSE_UPDATED
} from '@hoc/withSocket';
import '../AudioPrompt/AudioPrompt.css';
import Layout from '@utils/Layout';
import Media, { IS_AUDIO_RECORDING_SUPPORTED } from '@utils/Media';
import Storage from '@utils/Storage';

import './AnnotationPrompt.css';

class Display extends Component {
  constructor(props) {
    super(props);

    const { persisted = { value: [] } } = this.props;

    this.state = {
      isReady: false,
      activePage: 0,
      responses: [],
      entries: {},
      value: persisted.value || []
    };

    this.created_at = new Date().toISOString();
    this.onClick = this.onClick.bind(this);
    this.onResponseReceived = this.onResponseReceived.bind(this);
    this.onPageChange = this.onPageChange.bind(this);
  }

  get isScenarioRun() {
    return window.location.pathname.includes('/run/');
  }

  async componentDidMount() {
    if (!this.isScenarioRun) {
      return;
    }
    let { onResponseChange, persisted = {}, responseId } = this.props;
    let { value = [] } = persisted;

    let activePage = null;
    let entries = {};

    if (!value.length) {
      const previous = await this.props.getResponse(
        this.props.run.id,
        responseId
      );

      if (previous && previous.response) {
        value = previous.response.value;
      }
    }

    const responses = [];

    for (let promptResponseId of this.props.prompts) {
      const previous = await this.props.getResponse(
        this.props.run.id,
        promptResponseId
      );
      responses.push(previous);

      if (Array.isArray(value)) {
        const entry = value.find((v, index) => {
          if (v && v.response.response_id === promptResponseId) {
            activePage = index + 1;
            return v;
          }
        });
        entries[promptResponseId] = entry || null;
      }
    }

    const missingAnnotation = responses.find((response, index) => {
      if (!entries[response.response_id]) {
        activePage = index + 1;
        return response;
      }
    });

    if (!missingAnnotation) {
      let name = responseId;
      onResponseChange({}, { name, value, isFulfilled: true });
    }

    this.props.socket.on(RUN_RESPONSE_CREATED, this.onResponseReceived);
    this.props.socket.on(RUN_RESPONSE_UPDATED, this.onResponseReceived);

    this.setState({
      isReady: true,
      activePage,
      entries,
      responses,
      value
    });
  }

  componentWillUnmount() {
    this.props.socket.off(RUN_RESPONSE_CREATED, this.onResponseReceived);
    this.props.socket.off(RUN_RESPONSE_UPDATED, this.onResponseReceived);
  }

  async onResponseReceived(record) {
    const response = await this.props.getResponse(
      this.props.run.id,
      record.response_id
    );

    const entries = this.state.entries;
    const responses = this.state.responses;
    const index = responses.findIndex(
      ({ response_id }) => response_id === record.response_id
    );

    if (index !== -1) {
      responses[index] = {
        ...response
      };
    } else {
      responses.push(response);
    }

    this.setState({
      responses,
      entries: {
        ...entries,
        [record.response_id]: null
      }
    });
  }

  onClick(event, { isFulfilled, component, name, value, response }) {
    const { question, responseId, run } = this.props;
    const { created_at } = this;
    const ended_at = new Date().toISOString();
    let responses = this.state.responses.slice();
    let activePage = this.state.activePage;
    let entries = {
      ...this.state.entries
    };

    const annotation = {
      question,
      value
    };

    const context = {
      responseId,
      annotation,
      component,
      response
    };

    this.props.saveRunEvent(ANSWER_ANNOTATION, context);

    entries[name] = context;

    const annotations = Object.values(entries);

    const data = {
      created_at,
      ended_at,
      isFulfilled,
      name: responseId,
      type,
      value: annotations
    };

    Storage.set(`run/${run.id}/${responseId}`, data);

    let nextActivePage = activePage;

    const missingAnnotation = responses.find((response, index) => {
      if (!entries[response.response_id]) {
        nextActivePage = index + 1;
        return response;
      }
    });

    if (!missingAnnotation) {
      // This is necessary to "release" the hold made by required
      // response prompts.
      data.isFulfilled = false;
      data.isFulfillmentOverride = true;
      this.props.onResponseChange(event, data);
    } else {
      activePage = nextActivePage;
    }

    this.setState({
      activePage,
      entries,
      value: annotations
    });
  }

  onPageChange(event, { activePage }) {
    this.setState({
      ...this.state,
      activePage
    });
  }

  render() {
    const { onClick, onPageChange } = this;
    const { activePage, entries, isReady, responses } = this.state;
    const {
      isEmbeddedInSVG,
      question,
      required,
      run,
      scenario: { slides }
    } = this.props;

    if (!this.isScenarioRun || isEmbeddedInSVG) {
      return (
        <Message
          floating
          content="Participant responses will appear here during scenario run."
        />
      );
    }

    if (!isReady) {
      return (
        <Fragment>
          <Segment style={{ height: '300px' }}>
            <Header as="h3" tabIndex="0">
              {question}
            </Header>
            <Loading />
          </Segment>
        </Fragment>
      );
    }

    const missingAnnotation = responses.find(response => {
      if (!entries[response.response_id]) {
        return response;
      }
    });

    const activeResponseIndex = activePage - 1;
    const responseToAnnotate = responses[activeResponseIndex];

    const isFulfilled = !missingAnnotation;
    const header = (
      <Header as="h3" tabIndex="0">
        {question} {required && <PromptRequiredLabel fulfilled={isFulfilled} />}
      </Header>
    );

    const slideAndComponentAssociatedWithPrompt = slides.reduce(
      (accum, slide) => {
        const component = slide.components.find(
          ({ responseId }) => responseId === responseToAnnotate.response_id
        );
        if (component) {
          return {
            slide,
            component
          };
        }
        return accum;
      },
      undefined
    );

    const existing = entries[responseToAnnotate.response_id];

    const sharedButtonProps = {
      component: slideAndComponentAssociatedWithPrompt.component,
      response: responseToAnnotate,
      name: responseToAnnotate.response_id,
      onClick
    };

    const yesButton = {
      ...sharedButtonProps,
      icon:
        existing && existing.annotation.value === 'Yes' ? 'checkmark' : null,
      style: {
        background: 'rgb(115, 181, 128)'
      },
      value: 'Yes',
      content: 'Yes'
    };

    const noButton = {
      ...sharedButtonProps,
      icon: existing && existing.annotation.value === 'No' ? 'checkmark' : null,
      style: {
        background: 'rgb(217, 33, 32)',
        color: '#fff'
      },
      value: 'No',
      content: 'No'
    };

    const emptyButton = {
      ...sharedButtonProps,
      icon:
        existing && existing.annotation.value === 'No data'
          ? 'checkmark'
          : null,
      value: 'No data',
      content: 'No data'
    };

    const answerRecorded = responseToAnnotate && responseToAnnotate.response;
    let answerContent = answerRecorded
      ? answerRecorded.transcript ||
        answerRecorded.content ||
        answerRecorded.value
      : null;

    let answerContentDisplay = (
      <p>
        <strong>{answerContent}</strong>
      </p>
    );

    if (
      answerRecorded &&
      answerRecorded.isAudioPlayback &&
      IS_AUDIO_RECORDING_SUPPORTED
    ) {
      const src = Media.fileToMediaURL(answerRecorded.value);
      const audioProps = {
        controlsList: 'nodownload',
        controls: true,
        src
      };

      answerContentDisplay = (
        <div>
          <AudioPlayer {...audioProps} />
          <Transcript
            responseId={responseToAnnotate.response_id}
            run={run}
            transcript={answerRecorded.transcript}
          />
        </div>
      );
    }

    const urlToSlideWithThisComponent = `${window.location.pathname.slice(
      0,
      window.location.pathname.lastIndexOf('/')
    )}/${slideAndComponentAssociatedWithPrompt.slide.slide_number}`;

    const buttonGroupProps = {
      fluid: true
    };

    if (Layout.isNotForMobile()) {
      buttonGroupProps.widths = 3;
    }

    return responseToAnnotate ? (
      <Fragment>
        <Segment attached="top" className="ap__border-bottom-zero">
          {header}
          <Card fluid>
            <Card.Content>
              <Card.Header>
                {slideAndComponentAssociatedWithPrompt.component.prompt}
              </Card.Header>
              <Card.Meta>
                (From slide #
                {slideAndComponentAssociatedWithPrompt.slide.slide_number})
              </Card.Meta>
              <Card.Description>
                You answered:
                {answerContentDisplay}
                {/*answerRecorded.isEmpty ? (
                  <Button
                    fluid
                    as="a"
                    content={`Return to slide #${slideAndComponentAssociatedWithPrompt.slide.slide_number} to respond to this prompt`}
                    href={urlToSlideWithThisComponent}
                  />
                ) : null*/}
              </Card.Description>
            </Card.Content>
            <Card.Content extra>
              <Button.Group {...buttonGroupProps}>
                <Button {...yesButton} />
                <Button.Or />
                <Button {...noButton} />
                <Button.Or />
                <Button {...emptyButton} />
              </Button.Group>
            </Card.Content>
          </Card>
        </Segment>
        <Pagination
          borderless
          attached="bottom"
          className="ap__border-top-zero"
          name="agents"
          activePage={activePage}
          siblingRange={1}
          boundaryRange={0}
          ellipsisItem={null}
          firstItem={null}
          lastItem={null}
          onPageChange={onPageChange}
          totalPages={responses.length}
        />
      </Fragment>
    ) : null;
  }
}

Display.defaultProps = {
  isEmbeddedInSVG: false
};

Display.propTypes = {
  isEmbeddedInSVG: PropTypes.bool,
  getResponse: PropTypes.func,
  chat: PropTypes.object,
  onResponseChange: PropTypes.func,
  persisted: PropTypes.object,
  prompts: PropTypes.array,
  question: PropTypes.any,
  required: PropTypes.bool,
  responseId: PropTypes.string,
  run: PropTypes.object,
  saveRunEvent: PropTypes.func,
  scenario: PropTypes.object,
  socket: PropTypes.object,
  type: PropTypes.oneOf([type])
};

const mapStateToProps = state => {
  const { chat, scenario, run, responsesById, user } = state;
  return { chat, scenario, run, responsesById, user };
};

const mapDispatchToProps = dispatch => ({
  getResponse: (...params) => dispatch(getResponse(...params))
});

export default withSocket(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Display)
);
