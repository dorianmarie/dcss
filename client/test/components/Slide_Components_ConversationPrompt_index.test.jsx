import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect
}));

import assert from 'assert';
import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import { mount, shallow } from 'enzyme';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {
  type,
  name,
  description
} from '@components/Slide/Components/ConversationPrompt/meta';
import Card from '@components/Slide/Components/ConversationPrompt/Card';
import Display from '@components/Slide/Components/ConversationPrompt/Display';
import Editor from '@components/Slide/Components/ConversationPrompt/Editor';
import * as component from '../../components/Slide/Components/ConversationPrompt/index.js';

let container = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  jest.restoreAllMocks();
});

beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);
});

afterEach(() => {
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

test('type', () => {
  expect(component.type).toBe(type);
});

test('name', () => {
  expect(component.name).toBe(name);
});

test('description', () => {
  expect(component.description).toBe(description);
});

test('Display', () => {
  expect(component.Display).toBe(Display);
});

test('Editor', () => {
  expect(component.Editor).toBe(Editor);
});

test('Card', async () => {
  expect(component.Card).toBe(Card);
  expect(snapshotter(mounter(component.Card))).toMatchSnapshot();
  expect(snapshotter(mounter(Card))).toMatchSnapshot();
  expect(snapshotter(mounter(Card))).toBe(snapshotter(mounter(component.Card)));

  const { Card: ComponentCard } = component;

  const ConnectedRoutedComponent = reduxer(ComponentCard);
  const { asFragment } = render(<ConnectedRoutedComponent />);
  await screen.getByLabelText(component.name);
  expect(asFragment()).toMatchSnapshot();
});

test('Render 1', async () => {
  const params = {};

  expect(component.defaultValue(params)).toMatchInlineSnapshot(`
    Object {
      "configuration": Object {
        "duration": 0,
        "end": 0,
        "kind": "whole",
        "start": 0,
      },
      "header": "",
      "id": "",
      "persona": null,
      "prompt": "",
      "recallId": "",
      "required": true,
      "responseId": undefined,
      "timeout": 0,
      "type": "ConversationPrompt",
      "url": "",
    }
  `);
});

test('Render 2', async () => {
  const params = { recallId: 'ABC' };

  expect(component.defaultValue(params)).toMatchInlineSnapshot(`
    Object {
      "configuration": Object {
        "duration": 0,
        "end": 0,
        "kind": "whole",
        "start": 0,
      },
      "header": "",
      "id": "",
      "persona": null,
      "prompt": "",
      "recallId": "",
      "required": true,
      "responseId": undefined,
      "timeout": 0,
      "type": "ConversationPrompt",
      "url": "",
    }
  `);
});

test('Render 3', async () => {
  const params = { recallId: 'ABC', responseId: 'XYZ' };

  expect(component.defaultValue(params)).toMatchInlineSnapshot(`
    Object {
      "configuration": Object {
        "duration": 0,
        "end": 0,
        "kind": "whole",
        "start": 0,
      },
      "header": "",
      "id": "",
      "persona": null,
      "prompt": "",
      "recallId": "",
      "required": true,
      "responseId": "XYZ",
      "timeout": 0,
      "type": "ConversationPrompt",
      "url": "",
    }
  `);
});

/*{INJECTION}*/
