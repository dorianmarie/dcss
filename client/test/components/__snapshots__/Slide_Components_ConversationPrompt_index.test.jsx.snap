// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card 1`] = `"<i aria-label=\\"Prompt: Conversation\\" class=\\"conversation icon\\"></i><span class=\\"ser__component-name\\">Prompt: Conversation</span>"`;

exports[`Card 2`] = `"<i aria-label=\\"Prompt: Conversation\\" class=\\"conversation icon\\"></i><span class=\\"ser__component-name\\">Prompt: Conversation</span>"`;

exports[`Card 3`] = `
<DocumentFragment>
  <i
    aria-label="Prompt: Conversation"
    class="conversation icon"
  />
  <span
    class="ser__component-name"
  >
    Prompt: Conversation
  </span>
</DocumentFragment>
`;
